decomp/semantics/uds.py

```diff
@@ -960,37 +1188,199 @@ def _add_edge_annotation(self, edge, attrs):
+    def _get_timestamp_from_document_name(document_name):
+        timestamp = re.search('\d{8}_?\d{6}', document_name)
+        return timestamp[0] if timestamp else None
+
+    def add_sentence_graphs(self, sentence_graphs: Dict[str, UDSSentenceGraph], 
+                                  sentence_ids: Dict[str, str]) -> None:
+        """Add additional sentences to a document
 
-        root_candidates = [node for (node, degree) in digraph.in_degree() if degree == 0]
-        if len(root_candidates) != 1:
-            raise Exception('no unique root of digraph {}'.format(digraph.name))
-        return root_candidates[0]
+        Parameters
+        ----------
+        sentence_graphs
+            a dictionary containing the sentence-level graphs
+            for the sentences in the document
+        sentence_ids
+            a dictionary containing the UD sentence IDs for each graph
+        name
+            identifier to append to the beginning of node ids
+        """
+        for gname, graph in sentence_graphs.items():
+            sentence_graphs[gname].sentence_id = sentence_ids[gname]
+            sentence_graphs[gname].document_id = self.name
+            self.sentence_graphs[gname] = graph
+            self.sentence_ids[gname] = sentence_ids[gname]
+            for node in graph.semantics_nodes:
+                semantics = {'graph': gname, 'node': node}
+                document_node_name = node.replace('semantics', 'document')
+                self.document_graph.graph.add_node(document_node_name,
+                            domain='document', type='relation',
+                            frompredpatt=False, semantics=semantics)
+
+    def add_annotation(self, node_attrs: Dict[str, Dict[str, Any]],
+                             edge_attrs: Dict[str, Dict[str, Any]]) -> None:
+        """Add node or edge annotations to the document-level graph
 
-    @memoized_property
-    def document_id(self) -> str:
-        """Document ID from Universal Dependencies"""
-        syntax_root = self.get_digraph_root(self.syntax_subgraph)
-        return self.syntax_nodes[syntax_root]['document_id']
+        Parameters
+        ----------
+        node_attrs
+            the node annotations to be added
+        edge_attrs
+            the edge annotations to be added
+        """
+        self.document_graph.add_annotation(node_attrs, edge_attrs)
 
-    @memoized_property
-    def sentence_id(self) -> str:
-        """Sentence ID from Universal Dependencies"""
-        syntax_root = self.get_digraph_root(self.syntax_subgraph)
-        return self.syntax_nodes[syntax_root]['sentence_id']
+    def semantics_node(self, document_node: str) -> Dict:
+        """Fetch the semantics node for a given document node
 
+        Parameters
+        ----------
+        document_node
+            the document domain node whose semantics node is to be
+            retrieved
+        """
+        node = document_node.replace('semantics', 'document')
```

Comment:
This looks like the wrong direction of replacement. But is there any way to avoid munging node ID strings like this? At the very least, I think it would be useful to store the node identifier as an attribute rather than doing a replace like this on-the-fly. I'm sure it works now, but exactly this sort of string-munging of node IDs caused me a lot of headaches when I changed aspects of the node identifier scheme, which is why I've tried to avoid it wherever possible. I don't think we want to rely on the structure of the node ID strings staying constant. 

Also, I think this method implies that we retain `pred` or `arg` as part of the document node string, correct? I think this is what we want, because it may be that we need two separate document nodes corresponding to the same head position (e.g. for copular predicates and embedding clauses, where an argument and a predicate are headed by the same syntax node), just as in the semantics.

Reply:
Looking back, this looks to be a copy-paste error from line 1354.

Reply:
Can't we just use the semantics node ID added at line 1357?

Reply:
I think I wasn't clear exactly on what you meant by munging node ID strings before, but that makes sense. I'll store the semantics node name as an attribute on the document node name.

In any case, yes—I've preserved `pred` and `arg` as part of the name.

Reply:
It's already stored at line 1357, right?

Reply:
Well, it's already stored in the `semantics` attribute of the document node, so I'm not sure why I was doing things this way at all...

Reply:
Right. That's what I meant.